File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        ss = [nums[0]]
        for i in range(1, len(nums)):
            lo, hi = 0, len(ss)
            while lo < hi:
                mid = lo + (hi - lo) // 2
                if ss[mid] >= nums[i]:
                    hi = mid
                else:
                    lo = mid + 1
            if lo == len(ss):
                ss.append(nums[i])
            else:
                ss[lo] = nums[i]
        return len(ss)

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        ArrayList<Integer> maxLIS = new ArrayList<>();
        for(int i = 0; i < nums.length; i++){
            int lo = 0, hi = maxLIS.size();
            while(lo < hi){
                int mid = lo + (hi - lo) / 2;
                if(maxLIS.get(mid) >= nums[i])
                    hi = mid;
                else
                    lo = mid + 1;
            }
            if(lo == maxLIS.size())
                maxLIS.add(nums[i]);
            else
                maxLIS.set(lo, nums[i]);
        }
        return maxLIS.size();
    }
}